compiler/rustc_lint/src/function_cast_as_integer.rs

```diff
@@ -0,0 +1,82 @@
+use rustc_hir as hir;
+use rustc_macros::{LintDiagnostic, Subdiagnostic};
+use rustc_middle::ty::{self, Ty};
+use rustc_session::{declare_lint, declare_lint_pass};
+use rustc_span::{BytePos, Span};
+
+use crate::{LateContext, LateLintPass};
+
+declare_lint! {
+    /// The lint detects cases where users cast a function into an integer.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// fn foo() {}
+    /// let x = foo as usize;
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// You should never cast a function directly into an integer but go through
+    /// a cast as `fn` first to make it obvious what's going on. It also allows
+    /// to prevent confusion with (associated) constants.
+    pub FUNCTION_CASTS_AS_INTEGER,
+    Warn,
```

Comment:
Clippy has [a few lints](https://rust-lang.github.io/rust-clippy/master/index.html#/fn_to_) for `fn` to integer casts. But they are all restriction or style lints in Clippy. Adding a warn-by-default lint about this to rustc might be a bit aggressive :thinking: